#429B update action to regenerate review assignments

nmadruga opened the pull request:
Partly fixes #429

### Changes
- When `generateReviewAssignment` runs now it can be either to **generate** reviewAssigmnents (as previously) or **update** existing reviewAssignments after permissionNames were joined/unjoined to reviewers (users).
- Correctly list the review assignments that were updated/created after granting the same permissionName required for an application/stage/level. And also start listing the review assigments (and `reviewAssignmentAssignerJoins` for assignments) 
that were deleted after revoking a permissionName that was required for an application/stage/level.
- New tests to cover new cases added to `plugins/action_generate_review_assignment_records/src/generateReviewAssignments.test.ts`

CarlosNZ commented:
Looks good, but because this action has incredibly convoluted code, it's hard to figure out if it's all working as in should. 

In particular, I'm wondering about this case, from our agreed solution:
> if there are some, then it's assumed that this is a "regeneration" and it should run for the current (i.e. highest) review stage/level and not the next one which would be the case when supplied with a reviewId (as in a review submission)

I can't figure out where this is being done. It looks like the "existingReviewAssignments" fetch is occuring after the stage/level calculation is being done, but wouldn't that be needed first for the above situation? My apologies if I'm just not understanding this correctly.

nmadruga commented:
> I can't figure out where this is being done. It looks like the "existingReviewAssignments" fetch is occurring after the stage/level calculation is being done, but wouldn't that be needed first for the above situation? My apologies if I'm just not understanding this correctly.

I think you are right Carl, I haven't actually added this code yet. I was working with the tests, and it seemed to be creating the new reviewAssignment on the current stage, but I wasn't sure how. I'll have a better look into this now.

nmadruga commented:
Ok - @CarlosNZ I have done the changes needed - I didn't have to delete ReviewAssignmentAssigner records in the end... When I deleted the ReviewAssignments these were also being removed already. So that was the problem.

I'm starting to work on a diagram, but it would be good to get this one tested today and an action added to the template to grantUserPermissions (running after the granted/revoked permissions) happened. Also for when an internal user is added. I can start on that if you would like.

CarlosNZ commented on plugins/action_generate_review_assignment_records/src/generateReviewAssignments.test.ts:
```diff
@@ -112,6 +112,7 @@ test('Test: Submit Application ID#4001 - Stage 1 (Last level)', () => {
         reviewAssignmentIds: [1005, 1062, 1063, 1064, 1065],
         reviewAssignmentAssignerJoins: [],
         reviewAssignmentAssignerJoinIds: [],
+        removedAssignmentIds: [],
```
Not really important, but I notice some of the existing tests have a "templateId" parameter, which is not required. Might be best to get rid of it?

nmadruga replied:
Hmm. Sure. I'll take it off now : )

CarlosNZ commented on plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts:
```diff
@@ -25,6 +26,7 @@ async function generateReviewAssignments({
   // Get application/reviewId from applicationData if not provided in parameters
   const applicationId = parameters?.applicationId ?? applicationData?.applicationId
   const reviewId = parameters?.reviewId ?? applicationData?.reviewData?.reviewId
+  const isRegenaration = parameters?.isRegenaration ?? false
```
Spelling: `isRegeneration`

nmadruga replied:
Thanks!

CarlosNZ commented:
![Screen Shot 2021-12-01 at 1 25 08 PM](https://user-images.githubusercontent.com/5456533/144149473-733ac9a4-9607-4614-8a2b-3d5301b9d332.png)

You were asking about the `isReview` parameter. It's for when you're submitting a reviewId but you want to generate assignments for the first level anyway (so it's acts like you're submitting an application).

Considering that, you can probably consolidate `isReview` and `isRegeneration` into one parameter, since they almost do the same thing. Will help when I can see the exact logic written out, but hopefully we can get rid of one of them.

CarlosNZ commented:
Okay, so I'm having trouble getting the re-generation of "current" level assignments to work.

Testing with "dev" snapshot, using "ReviewTest" template. Got an application to Stage 2, level 2, all assignments created correctly along the way.

I can delete some of the Level 1 assignments and they get regenerated fine. But I can't get any of the level 2 assignments to regenerate, no matter what I set `isRegeration` to.

Using `/run-action` endpoint with the following parameters:
```
{
    "actionCode": "generateReviewAssignments",
    "applicationId": 4019,
    "parameters": {
        "applicationId": 4019,
        "isRegenaration": false
    }
}
```

It doesn't matter if I switch `isRegenation` to true or false, it only re-creates the Level 1 assignments.

Am I missing something?

nmadruga commented:
@CarlosNZ Phew, finally done - and docs updated. Please have another look and see what you think.

Basically changed to what was agreed yesterday: 
- Removed `isReview` flag no longer required
- Rewritten code to make it clearar so sorting what **type** of generation at the start and calling helper functions to deal with each case - required since now one of the cases would be running a loop.
- New logic for `isRegeneration` to generate reviewAssignments for **all** levels from 1 to current (based on existing reviewAssignments) to be creating or deleting the required as matching with reviewers with permissions to REVIEW.

CarlosNZ commented on plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts:
```diff
@@ -19,168 +21,299 @@ async function generateReviewAssignments({
-        Object.values(reviewAssignments) as ReviewAssignment[]
-      )
+  // Create array with levels [1,2..,N]
+  const arrayLevels = Array.from({ length: currentReviewLevel }, (v, k) => k + 1)
```
Fancy! 😉

CarlosNZ reviewed (changes requested):
Great -- that's a lot of work you did yesterday, but it's looking really good and everything seems to work in a way that makes sense (to me at least!) but without being overly complicated.

Couple of very minor suggestions.

We will need to keep an eye on this in the next couple of weeks, cos it is pretty complicated and there might still be unexpected issues, but I think it's pretty much good to go!

CarlosNZ commented:
Oh, one more thing I think this might need:

You raised a good point in [this comment](https://github.com/openmsupply/application-manager-server/pull/633#discussion_r759794184) about applications in "DRAFT" status. However if we include the "DRAFT" ones, then this Action will generate review_assignments for applications that haven't been submitted yet.

So what we should probably do is, in _this_ Action, do a check right at the top:
- if status == DRAFT AND there are no existing review_assignments, then exit immediately.

Do you think that will take care of it, or can you think of a better way? I could put this check in my Action and so we won't even pass those applications to this action, but I think it would be safer in here, as even if somehow a "DRAFT" application gets passed into this action accidentally, it still won't create invalid review assignments.

(Just looking at your code, you could move the `currentReviewLevel = (await db.getLastReviewLevel` up  to the top of the main function in order to do this check -- we already know status in applicationData -- then pass in the "current/highest" level into `generateForAllLevelsUntilCurrentLevel` as a parameter?)

nmadruga commented on plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts:
```diff
@@ -19,168 +21,299 @@ async function generateReviewAssignments({
+        numReviewLevels
       )
+    )
+  ).then((values) => values.map((reviewResult) => result.output.levels.push(reviewResult)))
```
Thanks! That's sounds better.

CarlosNZ replied:
64a2555c

CarlosNZ commented on plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts:
```diff
@@ -19,168 +21,299 @@ async function generateReviewAssignments({
-        stageNumber,
-        timeStageCreated: stageHistoryTimeCreated,
-      })
+const generateForAllLevelsUntilCurrentLevel = async (
```
1fbdfbf7

nmadruga commented:
@CarlosNZ merging this now and going to merge `develop` into `main`.
